spring-rabbit/src/main/java/org/springframework/amqp/rabbit/annotation/MultiRabbitListenerAnnotationBeanPostProcessor.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2002-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.amqp.rabbit.annotation;
+
+import java.lang.reflect.Method;
+
+import org.springframework.amqp.core.AbstractExchange;
+import org.springframework.amqp.core.Binding;
+import org.springframework.amqp.core.Declarable;
+import org.springframework.amqp.core.Queue;
+import org.springframework.amqp.rabbit.core.RabbitAdmin;
+import org.springframework.beans.factory.BeanFactory;
+import org.springframework.beans.factory.BeanFactoryAware;
+import org.springframework.beans.factory.NoSuchBeanDefinitionException;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.ApplicationContextAware;
+import org.springframework.lang.NonNull;
+
+/**
+ * An extension of {@link RabbitListenerAnnotationBeanPostProcessor} that associates the proper RabbitAdmin
+ * to the beans of Exchanges, Queues, and Bindings after they are created.
+ * <p>
+ * This processing restricts the {@link RabbitAdmin} according to the related configuration, preventing the server
+ * from automatic binding non-related structures.
+ *
+ * @author Wander Costa
+ * @see RabbitListenerAnnotationBeanPostProcessor
+ */
+public class MultiRabbitListenerAnnotationBeanPostProcessor extends RabbitListenerAnnotationBeanPostProcessor
+		implements ApplicationContextAware, BeanFactoryAware {
+
+	private ApplicationContext applicationContext;
+	private BeanFactory beanFactory;
+
+	@Override
+	protected void processAmqpListener(RabbitListener rabbitListener, Method method, Object bean, String beanName) {
+		super.processAmqpListener(rabbitListener, method, bean, beanName);
+		enhanceBeansWithReferenceToRabbitAdmin(rabbitListener);
+	}
+
+	/**
+	 * Enhance beans with related RabbitAdmin, so as to be filtered when being processed by the RabbitAdmin.
+	 *
+	 * @param rabbitListener the RabbitListener to enhance its bean.
+	 */
+	private void enhanceBeansWithReferenceToRabbitAdmin(RabbitListener rabbitListener) {
+		RabbitAdmin rabbitAdmin = getRabbitAdminBean(rabbitListener);
+		// Enhance Exchanges
+		this.applicationContext.getBeansOfType(AbstractExchange.class).values().stream().filter(this::isNotProcessed)
+				.forEach(exchange -> exchange.setAdminsThatShouldDeclare(rabbitAdmin));
```

Comment:
No. The declarables (`AbstractExchange`, `Queue` and `Binding`) are not declared on all servers.

That's the reason for the `isNotProcessed()` method.
Once the declarable beans are created by BPP for one `@RabbitListener`, they will not have any `RabbitAdmin` yet associated.
Then, they are enhanced with the corresponding `RabbitAdmin`.

Then, when BPP processes the 2nd, 3rd, etc `@RabbitListener`s, the previous declarables will be filtered out and not enhanced, because they already have a `RabbitAdmin` associated to them.

This way, only the last declarables created are associated to the current `RabbitAdmin`.

Reply:
Oh; now I see the reason for my confusion; this is a bit brittle - you are using the fact that the BPP "just" added the beans  for the current `@RabbitListener.queueBindings()` and the beans for other `@QueueBinding`s don't exist yet.

The problem with this approach (using `getBeansOfType()`) is that all stand-alone queue/exchange/binding `@Bean`s will be declared by the first admin.

We could use the names from the `@QueueBinding`s and do `getBean()` for each individually to avoid picking up other unrelated beans.

However, this is complicated by the fact that the bean names are `foo.n` (where n is incremented with a common integer) so there is no way to know which beans were registered for the current annotation.

I can see a couple of possible solutions - 

- change `processAmqpListener()` to return `Collection<Declarable>` containing any beans that were registered for this listener in the super class
- add a new `@RabbitListener` field to `AbstractDeclarable` and use that as the `.filter`.